### PR TITLE
prov/efa: Directly use ibv_cq_ex in rxr_ep 

### PR DESCRIPTION
**Pull request by sunkuamzn**
rxr_ep now has a new member ibv_cq_ex which is polled directly in
rxr_ep_progress_internal

Signed-off-by: Sai Sunku <sunkusa@amazon.com>

### CODE REVIEW

**Comment by sunkuamzn** [comment]
Will fix the space/tab in `struct ibv_cq_ex *ibv_cq_ex;` in next revision

File: prov/efa/src/rxr/rxr.h [hunk 1]
```diff
@@ -291,7 +291,7 @@ struct rxr_ep {
 
 	/* core provider fid */
 	struct fid_ep *rdm_ep;
-	struct fid_cq *rdm_cq;
+    struct ibv_cq_ex *ibv_cq_ex;
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr.h:294]: use tab other than space.

File: prov/efa/src/rxr/rxr_ep.c [hunk 2]
```diff
@@ -743,11 +743,8 @@ static int rxr_ep_close(struct fid *fid)
 		retv = ret;
 	}
 
-	ret = fi_close(&rxr_ep->rdm_cq->fid);
-	if (ret) {
-		FI_WARN(&rxr_prov, FI_LOG_EP_CTRL, "Unable to close msg CQ\n");
-		retv = ret;
-	}
+	struct ibv_cq *ibv_cq = ibv_cq_ex_to_cq(rxr_ep->ibv_cq_ex);
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_ep.c:746]: declare the variable at the beginning of the function.

File: prov/efa/src/rxr/rxr_ep.c [hunk 3]
```diff
@@ -743,11 +743,8 @@ static int rxr_ep_close(struct fid *fid)
...
+	ret = -ibv_destroy_cq(ibv_cq);
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_ep.c:747]: check `ret` and print warning if it is not 0.

File: prov/efa/src/rxr/rxr_ep.c [hunk 4]
```diff
@@ -2484,6 +2479,66 @@ int rxr_ep_alloc_app_device_info(struct fi_info **app_device_info,
...
 
+/**
+ *
+ * @brief Create ibv_cq_ex by calling ibv_create_cq_ex
+ *
+ * @param[in] cq Pointer to the efa_cq.
+ * @param[in] attr Pointer to fi_cq_attr.
+ * @param[out] Return pointer to ibv_cq_ex if successful, otherwise NULL
+ */
+static inline struct ibv_cq_ex *rxr_ep_create_ibv_cq_ex(struct rxr_ep *ep, struct fi_cq_attr *attr) {
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_ep.c:2490]: `{` need to be the next line.

File: prov/efa/src/rxr/rxr_ep.c [hunk 5]
```diff
@@ -2484,6 +2479,66 @@ int rxr_ep_alloc_app_device_info(struct fi_info **app_device_info,
...
+		.wc_flags = IBV_WC_STANDARD_FLAGS,
+		.comp_mask = 0,
+	};
+
+	struct efa_domain *efa_domain = rxr_ep_domain(ep);
+	return ibv_create_cq_ex(efa_domain->device->ibv_ctx, &init_attr_ex);
+}
+
+static inline int rxr_ep_set_ibv_cq_ex(struct rxr_ep *ep, struct fi_cq_attr *attr) {
+	if (ep->ibv_cq_ex) {
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_ep.c:2508]: I think this function can be folded into `rxr_endpoint`.

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c:2508]: Not sure this check is useful.

I think an assertion is good enough.

File: prov/efa/src/rxr/rxr_ep.c [hunk 6]
```diff
@@ -2484,6 +2479,66 @@ int rxr_ep_alloc_app_device_info(struct fi_info **app_device_info,
...
+}
+
+static int efa_ep_bind_rdm(struct rxr_ep *rxr_ep, struct ibv_cq_ex *ibv_cq_ex)
+{
+	struct efa_ep* efa_ep = container_of(rxr_ep->rdm_ep, struct efa_ep, util_ep.ep_fid);
+
+	efa_ep->scq = malloc(sizeof(struct efa_cq));
+	if (!efa_ep->scq)
+		return -FI_ENOMEM;
+    efa_ep->scq->ibv_cq_ex = ibv_cq_ex;
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_ep.c:2530]: indentation seems off.

File: prov/efa/src/rxr/rxr_ep.c [hunk 7]
```diff
@@ -2641,17 +2692,10 @@ int rxr_endpoint(struct fid_domain *domain, struct fi_info *info,
 				fi_strerror(-retv));
 	}
 err_close_core_cq:
-	retv = fi_close(&rxr_ep->rdm_cq->fid);
+	retv = -ibv_destroy_cq(ibv_cq_ex_to_cq(rxr_ep->ibv_cq_ex));
 	if (retv)
 		FI_WARN(&rxr_prov, FI_LOG_CQ, "Unable to close cq: %s\n",
 			fi_strerror(-retv));
-err_close_shm_ep:
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_ep.c:2648]: Why remove this part?

**Review by wzamazon — CHANGES_REQUESTED** [review]

File: prov/efa/src/rxr/rxr_ep.c (hunk 7 above)
Comment (sunkuamzn) [reply prov/efa/src/rxr/rxr_ep.c:2648]: It is now unused. It was only used on line 2601 of the old file and I removed that.

File: prov/efa/src/rxr/rxr_ep.c (hunk 6 above)
Comment (sunkuamzn) [reply prov/efa/src/rxr/rxr_ep.c:2530]: ACK

File: prov/efa/src/rxr/rxr_ep.c (hunk 5 above)
Comment (sunkuamzn) [reply prov/efa/src/rxr/rxr_ep.c:2508]: ACK

File: prov/efa/src/rxr/rxr_ep.c (hunk 3 above)
Comment (sunkuamzn) [reply prov/efa/src/rxr/rxr_ep.c:747]: will fix

**Review by shijin-aws — COMMENTED** [review]
Add prov/efa: as a prefix of your commit message

**Comment by wzamazon** [comment]
The commit message need some adjustment.

Generally speaking, the message should be about what it does not what it don't do.

e.g. instead of "Don't use efa_cq in rxr_ep", I think it is better to say "use ibv_cq_ex directly in rxr_ep".

File: prov/efa/src/rxr/rxr_ep.c [hunk 8]
```diff
@@ -2484,6 +2479,66 @@ int rxr_ep_alloc_app_device_info(struct fi_info **app_device_info,
...
+	efa_ep->scq = malloc(sizeof(struct efa_cq));
+	if (!efa_ep->scq)
+		return -FI_ENOMEM;
+    efa_ep->scq->ibv_cq_ex = ibv_cq_ex;
+	efa_ep->scq->domain = rxr_ep_domain(rxr_ep);
+
+	efa_ep->rcq = malloc(sizeof(struct efa_cq));
+	if (!efa_ep->rcq)
+		return -FI_ENOMEM;
+    efa_ep->rcq->ibv_cq_ex = ibv_cq_ex;
```
Comment (shijin-aws) [inline prov/efa/src/rxr/rxr_ep.c:2536]: another indent off here.

File: prov/efa/src/rxr/rxr_ep.c (hunk 7 above)
Comment (shijin-aws) [reply prov/efa/src/rxr/rxr_ep.c:2648]: But your `rxr_ep_set_ibv_cq_ex(rxr_ep, &cq_attr);` did not clean shm resource in error path? You actually change the behavior of the code?

Comment (sunkuamzn) [reply prov/efa/src/rxr/rxr_ep.c:2648]: fixed!

File: prov/efa/src/rxr/rxr_ep.c [hunk 9]
```diff
@@ -2596,13 +2643,16 @@ int rxr_endpoint(struct fid_domain *domain, struct fi_info *info,
...
-	if (ret)
-		goto err_close_shm_ep;
+	assert(!(rxr_ep->ibv_cq_ex));
+
+	rxr_ep->ibv_cq_ex = rxr_ep_create_ibv_cq_ex(rxr_ep, &cq_attr);
 
-	ret = fi_ep_bind(rxr_ep->rdm_ep, &rxr_ep->rdm_cq->fid,
-			 FI_TRANSMIT | FI_RECV);
+	if (!rxr_ep->ibv_cq_ex) {
+		EFA_WARN(FI_LOG_CQ, "Unable to create extended CQ: %s\n", strerror(errno));
```
Comment (shijin-aws) [inline prov/efa/src/rxr/rxr_ep.c:2651]: do you need to keep the `goto err_close_shm_ep` in the error path?

Comment (sunkuamzn) [reply prov/efa/src/rxr/rxr_ep.c:2651]: I think I understand now. does this look good?

File: prov/efa/src/rxr/rxr_ep.c [hunk 10]
```diff
@@ -743,11 +744,12 @@ static int rxr_ep_close(struct fid *fid)
...
 
-	ret = fi_close(&rxr_ep->rdm_cq->fid);
-	if (ret) {
-		FI_WARN(&rxr_prov, FI_LOG_EP_CTRL, "Unable to close msg CQ\n");
+	ibv_cq = ibv_cq_ex_to_cq(rxr_ep->ibv_cq_ex);
+
+	ret = -ibv_destroy_cq(ibv_cq);
+	if (ret)
+		FI_WARN(&rxr_prov, FI_LOG_EP_CTRL, "Unable to close ibv_cq_ex\n");
 		retv = ret;
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_ep.c:752]: Don't you need `{` here?

File: prov/efa/src/rxr/rxr_ep.c [hunk 11]
```diff
@@ -2484,6 +2485,52 @@ int rxr_ep_alloc_app_device_info(struct fi_info **app_device_info,
...
+		.wc_flags = IBV_WC_STANDARD_FLAGS,
+		.comp_mask = 0,
+	};
+
+	struct efa_domain *efa_domain = rxr_ep_domain(ep);
+
+	return ibv_create_cq_ex(efa_domain->device->ibv_ctx, &init_attr_ex);
+}
+
+static int efa_ep_bind_rdm(struct rxr_ep *rxr_ep, struct ibv_cq_ex *ibv_cq_ex)
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_ep.c:2515]: add a doxygen comment for this function too.

**Review by wzamazon — CHANGES_REQUESTED** [review]
In the commit message,
"Now rxr_ep has a new member xxx"

I think you meant

"This patch introduced a new member xxx to rxr_ep"

File: prov/efa/src/rxr/rxr_ep.c (hunk 10 above)
Comment (sunkuamzn) [reply prov/efa/src/rxr/rxr_ep.c:752]: thanks, will fix

File: prov/efa/src/rxr/rxr_ep.c (hunk 11 above)
Comment (sunkuamzn) [reply prov/efa/src/rxr/rxr_ep.c:2515]: I thought this function is temporary and will get removed when we separate endpoints. But I can add comments.

File: prov/efa/src/efa_ep.c [hunk 12]
```diff
@@ -251,6 +251,12 @@ static int efa_ep_close(fid_t fid)
 
 	ep = container_of(fid, struct efa_ep, util_ep.ep_fid.fid);
 
+	if (ep->scq)
+		free(ep->scq);
+
+	if (ep->rcq)
+		free(ep->rcq);
```
Comment (sunkuamzn) [inline prov/efa/src/efa_ep.c:258]: Need to free after malloc in `efa_ep_bind_rdm`

File: prov/efa/src/rxr/rxr_ep.c (hunk 11 above)
Comment (sunkuamzn) [reply prov/efa/src/rxr/rxr_ep.c:2522]: Should this function signature be `efa_ep_bind_rdm(struct rxr_ep *rxr_ep)` and call `ibv_cq_ex *ibv_cq_ex = rxr_ep->ibv_cq_ex` inside the function?

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c:2522]: Yes. I think so. also the function name need some adjustment, maybe `rxr_ep_setup_device_cq`?

Comment (sunkuamzn) [reply prov/efa/src/rxr/rxr_ep.c]: fixed

File: prov/efa/src/rxr/rxr_ep.c [hunk 13]
```diff
@@ -2596,13 +2651,17 @@ int rxr_endpoint(struct fid_domain *domain, struct fi_info *info,
 	rxr_ep->efa_outstanding_tx_ops = 0;
 	rxr_ep->shm_outstanding_tx_ops = 0;
 
-	ret = efa_cq_open(&efa_domain->util_domain.domain_fid, &cq_attr,
-			  &rxr_ep->rdm_cq, rxr_ep);
-	if (ret)
+	assert(!(rxr_ep->ibv_cq_ex));
```
Comment (shijin-aws) [inline prov/efa/src/rxr/rxr_ep.c:2654]: do you really need extra `()` around `rxr_ep->ibv_cq_ex`? especially you did not have it for L2658

File: prov/efa/src/rxr/rxr_ep.c [hunk 14]
```diff
@@ -2484,6 +2486,59 @@ int rxr_ep_alloc_app_device_info(struct fi_info **app_device_info,
...
+	struct ibv_cq_ex *ibv_cq_ex = rxr_ep->ibv_cq_ex;
+
+	efa_ep->scq = malloc(sizeof(struct efa_cq));
+	if (!efa_ep->scq)
+		return -FI_ENOMEM;
+	efa_ep->scq->ibv_cq_ex = ibv_cq_ex;
+	efa_ep->scq->domain = rxr_ep_domain(rxr_ep);
+
+	efa_ep->rcq = malloc(sizeof(struct efa_cq));
+	if (!efa_ep->rcq)
```
Comment (shijin-aws) [inline prov/efa/src/rxr/rxr_ep.c:2534]: when rcq malloc failed, do you need to free the `scq` ?

File: prov/efa/src/efa_ep.c (hunk 12 above)
Comment (wzamazon) [reply prov/efa/src/efa_ep.c:258]: I do not think you want to call `free` here. This function is used for both rdm and dgram, and for dgram `ep-scq` and `ep->rcq` is not allocated, so should not be freed. I am surprised that `fi_dgram_pingpong` did not fail.

File: prov/efa/src/rxr/rxr_ep.c [hunk 15]
```diff
@@ -743,9 +744,11 @@ static int rxr_ep_close(struct fid *fid)
 		retv = ret;
 	}
 
-	ret = fi_close(&rxr_ep->rdm_cq->fid);
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_ep.c:746]: I think the proper place to release the `scq` and `rcq` allocated in `rxr_ep_setup_device_cq` is here. Maybe introduce another function like `rxr_ep_release_device_cq` to do that.

**Review by wzamazon — CHANGES_REQUESTED** [review]

File: prov/efa/src/rxr/rxr_ep.c (hunk 14 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c:2534]: I think you do not need to allocate `rcq` here. set `efa_ep->rcq` to `efa_ep->scq` should work.

**Comment by wzamazon** [comment]
AWS CI failure is because of `fi_dgram_pingpong` failure, which is real. I am surprised that it passed on alinux2.

File: prov/efa/src/efa_ep.c (hunk 12 above)
Comment (sunkuamzn) [reply prov/efa/src/efa_ep.c]: fixed

File: prov/efa/src/rxr/rxr_ep.c (hunk 15 above)
Comment (sunkuamzn) [reply prov/efa/src/rxr/rxr_ep.c:746]: fixed

File: prov/efa/src/rxr/rxr_ep.c (hunk 14 above)
Comment (sunkuamzn) [reply prov/efa/src/rxr/rxr_ep.c]: fixed

File: prov/efa/src/rxr/rxr_ep.c (hunk 13 above)
Comment (sunkuamzn) [reply prov/efa/src/rxr/rxr_ep.c]: fixed

File: prov/efa/src/rxr/rxr_ep.c [hunk 16]
```diff
@@ -728,6 +728,25 @@ void rxr_ep_wait_send(struct rxr_ep *rxr_ep)
 	ofi_mutex_unlock(&rxr_ep->util_ep.lock);
 }
 
+/**
+ * @brief Destroy the ibv_cq_ex in rxr_ep
+ * 	Will be removed when rxr_ep and efa_ep are fully separated
+ *
+ * @param[in] rxr_ep Pointer to the rxr_ep.
+ * @param[in] ibv_cq_ex Pointer to ibv_cq_ex.
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_ep.c:736]: This param was not used.
also add a line for return value.

File: prov/efa/src/rxr/rxr_ep.c [hunk 17]
```diff
@@ -728,6 +728,25 @@ void rxr_ep_wait_send(struct rxr_ep *rxr_ep)
...
+ */
+static int rxr_ep_release_device_cq(struct rxr_ep *rxr_ep)
+{
+	struct efa_ep *efa_ep = container_of(rxr_ep->rdm_ep, struct efa_ep, util_ep.ep_fid);
+
+	if (efa_ep->scq)
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_ep.c:742]: I think we always allocate `efa_ep->scq`. Right?, so this should be an assertion.

File: prov/efa/src/rxr/rxr_ep.c [hunk 18]
```diff
@@ -728,6 +728,25 @@ void rxr_ep_wait_send(struct rxr_ep *rxr_ep)
...
+		free(efa_ep->scq);
+
+	struct ibv_cq *ibv_cq = ibv_cq_ex_to_cq(rxr_ep->ibv_cq_ex);
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_ep.c:745]: Declaration of variable should be at the beginning of the function

File: prov/efa/src/rxr/rxr_ep.c [hunk 19]
```diff
@@ -2596,13 +2663,17 @@ int rxr_endpoint(struct fid_domain *domain, struct fi_info *info,
 	rxr_ep->efa_outstanding_tx_ops = 0;
 	rxr_ep->shm_outstanding_tx_ops = 0;
 
-	ret = efa_cq_open(&efa_domain->util_domain.domain_fid, &cq_attr,
-			  &rxr_ep->rdm_cq, rxr_ep);
-	if (ret)
+	assert(!rxr_ep->ibv_cq_ex);
+
+	rxr_ep->ibv_cq_ex = rxr_ep_create_ibv_cq_ex(rxr_ep, &cq_attr);
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_ep.c:2668]: I think it is probably better that this part is moved into `rxr_ep_setup_device_cq`.

File: prov/efa/src/rxr/rxr_ep.c [hunk 20]
```diff
@@ -2484,6 +2502,55 @@ int rxr_ep_alloc_app_device_info(struct fi_info **app_device_info,
 	return 0;
 }
 
+/**
+ * @brief Create ibv_cq_ex by calling ibv_create_cq_ex
+ *
+ * @param[in] cq Pointer to the efa_cq.
+ * @param[in] attr Pointer to fi_cq_attr.
+ * @param[out] Return pointer to ibv_cq_ex if successful, otherwise NULL
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_ep.c:2510]: This should be `@return`.

`@param[out]` is for an argument

**Review by wzamazon — CHANGES_REQUESTED** [review]
Some minor comments.

File: prov/efa/src/rxr/rxr_ep.c (hunk 19 above)
Comment (sunkuamzn) [reply prov/efa/src/rxr/rxr_ep.c:2668]: I need the `err_close_shm_ep` label. This function won't be around in a few weeks anyway, so it doesn't have to be perfect?

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c:2668]: I see. Then no need to bother.

**Review by wzamazon — APPROVED** [review]

**Comment by shijin-aws** [comment]
may I know what caused earlier `fi_dgram_pingpong` failure and how it was fixed?

**Comment by sunkuamzn** [comment]
> may I know what caused earlier `fi_dgram_pingpong` failure and how it was fixed?

It was caused a double free because I called `free(efa->scq)` in `efa_ep_close`. Now, I'm calling `free(efa->scq)` in `rxr_ep_release_device_cq` which is only called by RDM and not by DGRAM

**Review by shijin-aws — APPROVED** [review]

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
Pure EFA change, so ignore intel CI failure.

**Pull request merged**